File: БР1.2/Гайдук Алина/labs/lab2/internal/infrastructure/middleware/service_token.go
package middleware

import (
	"crypto/subtle"
	"net/http"
	"strings"

	"recipehub/internal/transport/http/response"
)

// ServiceToken protects internal service-to-service routes.
type ServiceToken struct {
	Token string
}

// Required rejects requests without a valid X-Service-Token header.
func (s ServiceToken) Required(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		got := strings.TrimSpace(r.Header.Get("X-Service-Token"))
		want := strings.TrimSpace(s.Token)
		if want == "" || subtle.ConstantTimeCompare([]byte(got), []byte(want)) != 1 {
			response.RespondError(w, http.StatusUnauthorized, "UNAUTHORIZED", "service token required")
			return
		}

		next.ServeHTTP(w, r)
	})
}

File: services/auth/src/middleware/serviceToken.ts
import { NextFunction, Request, Response } from "express";
import { env } from "../config/env";
import { unauthorized } from "../utils/errors";

export function serviceTokenMiddleware(
  req: Request,
  _res: Response,
  next: NextFunction,
): void {
  const token = req.headers["x-service-token"];
  if (token !== env.serviceToken) {
    return next(unauthorized());
  }
  next();
}
